docs/source/examples/ldaptor_with_upn_bind.py

```diff
@@ -20,78 +20,44 @@
 
 from ldaptor import interfaces
 from ldaptor.protocols import pureldap
-from ldaptor.protocols.ldap import distinguishedname, ldaperrors
 from twisted.internet import defer
 from ldaptor.protocols.ldap.ldapserver import LDAPServer
 
 
-class LDAPServerWithUPNBind(LDAPServer):
+class LDAPServerWithUPNBind(LDAPServer, object):
     """
     An LDAP server which support BIND using UPN similar to AD.
     """
-    _loginAttribute = b'userPrincipalName'
 
-    def handle_LDAPBindRequest(self, request, controls, reply):
-        if request.version != 3:
-            raise ldaperrors.LDAPProtocolError(
-                'Version %u not supported' % request.version)
+    _loginAttribute = b"userPrincipalName"
 
-        self.checkControls(controls)
-
-        if request.dn == b'':
-            # anonymous bind
-            self.boundUser = None
-            return pureldap.LDAPBindResponse(resultCode=0)
-
-        root = interfaces.IConnectedLDAPEntry(self.factory)
-
-        def _gotUPNResult(results):
-            if len(results) != 1:
-                # Not exactly one result, so this might not be an UNP.
-                return distinguishedname.DistinguishedName(request.dn)
-
-            # A single result, so the UPN might exist.
-            return results[0].dn
-
-        if b'@' in request.dn and b',' not in request.dn:
+    @defer.inlineCallbacks
+    def handle_LDAPBindRequest(self, request, *args, **kwargs):
+        @defer.inlineCallbacks
+        def _request():
+            if not (b"@" in request.dn and b"," not in request.dn):
+                defer.returnValue(request)
```

Comment:
Maybe something like this

```suggestion
        def _resolveUPNBindDN():
           """
           See if we have an MS UPN style BIND DN request and resolve it to the
           actual LDAP BindDN.
           """
            if b"@" not in request.dn or b"," in request.dn):
                # This is not an UPN bind request
                defer.returnValue(request)
```